File: go-pure-impl/registry.go
package main

// OrderRegistry holds flat per-order metadata indexed by OrderID.
type OrderRegistry struct {
	Meta [NumOrders]OrderMeta
}

func NewOrderRegistry() *OrderRegistry { return &OrderRegistry{} }

func (r *OrderRegistry) Upsert(m OrderMeta) {
	r.Meta[m.OrderID] = m
}

func (r *OrderRegistry) Deactivate(orderID uint32) {
	r.Meta[orderID].Active = false
}

// EvictOrder deactivates an order and cleans up its index entries and PPSMatrix slots.
// Call this on Kafka order-removed events.
func (r *OrderRegistry) EvictOrder(orderID uint32, indexes *InvertedIndexes, matrix *PPSMatrix) {
	if int(orderID) >= NumOrders || !r.Meta[orderID].Active {
		return
	}
	r.Meta[orderID].Active = false
	r.Meta[orderID].InsertedAtSecs = 0
	indexes.RemoveOrder(orderID)
	matrix.EvictOrder(orderID)
}

// PurgeExpired scans all orders and evicts those whose TTL has expired.
// Call at the start of each compute cycle.
func (r *OrderRegistry) PurgeExpired(nowSecs int64, indexes *InvertedIndexes, matrix *PPSMatrix) {
	for i := 0; i < NumOrders; i++ {
		m := &r.Meta[i]
		if m.Active && m.InsertedAtSecs > 0 && (nowSecs-m.InsertedAtSecs) > OrderTTLSecs {
			orderID := m.OrderID
			m.Active = false
			m.InsertedAtSecs = 0
			indexes.RemoveOrder(orderID)
			matrix.EvictOrder(orderID)
		}
	}
}

// InvertedIndexes maps TPID and BinTag IDs to sets of order IDs.
type InvertedIndexes struct {
	TPIDToOrders   map[uint64][]uint32
	BinTagToOrders map[uint32][]uint32
	reverseTPID    map[uint32][]uint64
	reverseBinTag  map[uint32][]uint32
}

func NewInvertedIndexes() *InvertedIndexes {
	return &InvertedIndexes{
		TPIDToOrders:   make(map[uint64][]uint32),
		BinTagToOrders: make(map[uint32][]uint32),
		reverseTPID:    make(map[uint32][]uint64),
		reverseBinTag:  make(map[uint32][]uint32),
	}
}

func (ix *InvertedIndexes) AddOrder(orderID uint32, tpids []uint64, bintagIDs []uint32) {
	ix.RemoveOrder(orderID)

	for _, tpid := range tpids {
		ix.TPIDToOrders[tpid] = append(ix.TPIDToOrders[tpid], orderID)
	}
	for _, btID := range bintagIDs {
		ix.BinTagToOrders[btID] = append(ix.BinTagToOrders[btID], orderID)
	}

	if len(tpids) > 0 {
		cp := make([]uint64, len(tpids))
		copy(cp, tpids)
		ix.reverseTPID[orderID] = cp
	}
	if len(bintagIDs) > 0 {
		cp := make([]uint32, len(bintagIDs))
		copy(cp, bintagIDs)
		ix.reverseBinTag[orderID] = cp
	}
}

func (ix *InvertedIndexes) RemoveOrder(orderID uint32) {
	for _, tpid := range ix.reverseTPID[orderID] {
		list := ix.TPIDToOrders[tpid]
		list = removeUint32(list, orderID)
		if len(list) == 0 {
			delete(ix.TPIDToOrders, tpid)
		} else {
			ix.TPIDToOrders[tpid] = list
		}
	}
	delete(ix.reverseTPID, orderID)

	for _, btID := range ix.reverseBinTag[orderID] {
		list := ix.BinTagToOrders[btID]
		list = removeUint32(list, orderID)
		if len(list) == 0 {
			delete(ix.BinTagToOrders, btID)
		} else {
			ix.BinTagToOrders[btID] = list
		}
	}
	delete(ix.reverseBinTag, orderID)
}

func removeUint32(s []uint32, v uint32) []uint32 {
	for i, x := range s {
		if x == v {
			return append(s[:i], s[i+1:]...)
		}
	}
	return s
}

File: rust-impl/src/registry.rs
use crate::types::{NUM_ORDERS, ORDER_TTL_SECS, OrderMeta, PPSMatrix};
use std::collections::HashMap;

// ─── OrderRegistry ───────────────────────────────────────────────────────────

pub struct OrderRegistry {
    pub meta: Vec<OrderMeta>,
}

impl OrderRegistry {
    pub fn new() -> Self {
        Self {
            meta: vec![OrderMeta::default(); NUM_ORDERS],
        }
    }

    pub fn upsert(&mut self, m: OrderMeta) {
        let idx = m.order_id as usize;
        if idx < NUM_ORDERS {
            self.meta[idx] = m;
        }
    }

    pub fn deactivate(&mut self, order_id: u32) {
        let idx = order_id as usize;
        if idx < NUM_ORDERS {
            self.meta[idx].active = false;
        }
    }

    /// Evict a single order: mark inactive, remove from inverted indexes,
    /// zero PPSMatrix slots. Used for Kafka order-removed events.
    pub fn evict_order(
        &mut self,
        order_id: u32,
        indexes: &mut InvertedIndexes,
        matrix: &mut PPSMatrix,
    ) {
        let idx = order_id as usize;
        if idx >= NUM_ORDERS || !self.meta[idx].active {
            return;
        }
        self.meta[idx].active = false;
        self.meta[idx].inserted_at_secs = 0;
        indexes.remove_order(order_id);
        matrix.evict_order(idx);
    }

    /// Scan and evict all orders whose TTL has expired.
    /// Call at the start of each compute cycle.
    pub fn purge_expired(
        &mut self,
        now_secs: i64,
        indexes: &mut InvertedIndexes,
        matrix: &mut PPSMatrix,
    ) {
        for idx in 0..NUM_ORDERS {
            // Read the fields we need before any mutation.
            let active = self.meta[idx].active;
            let inserted_at = self.meta[idx].inserted_at_secs;
            let order_id = self.meta[idx].order_id;

            if active && inserted_at > 0 && (now_secs - inserted_at) > ORDER_TTL_SECS {
                self.meta[idx].active = false;
                self.meta[idx].inserted_at_secs = 0;
                indexes.remove_order(order_id);
                matrix.evict_order(idx);
            }
        }
    }
}

impl Default for OrderRegistry {
    fn default() -> Self {
        Self::new()
    }
}

// ─── InvertedIndexes ─────────────────────────────────────────────────────────

pub struct InvertedIndexes {
    pub tpid_to_orders: HashMap<u64, Vec<u32>>,
    pub bintag_to_orders: HashMap<u32, Vec<u32>>,
    order_tpids: HashMap<u32, Vec<u64>>,
    order_bintags: HashMap<u32, Vec<u32>>,
}

impl InvertedIndexes {
    pub fn new() -> Self {
        Self {
            tpid_to_orders: HashMap::new(),
            bintag_to_orders: HashMap::new(),
            order_tpids: HashMap::new(),
            order_bintags: HashMap::new(),
        }
    }

    pub fn add_order(&mut self, order_id: u32, tpids: &[u64], bintag_ids: &[u32]) {
        self.remove_order(order_id);

        for &tpid in tpids {
            self.tpid_to_orders.entry(tpid).or_default().push(order_id);
        }
        for &bt in bintag_ids {
            self.bintag_to_orders.entry(bt).or_default().push(order_id);
        }

        self.order_tpids.insert(order_id, tpids.to_vec());
        self.order_bintags.insert(order_id, bintag_ids.to_vec());
    }

    pub fn remove_order(&mut self, order_id: u32) {
        if let Some(tpids) = self.order_tpids.remove(&order_id) {
            for tpid in tpids {
                if let Some(v) = self.tpid_to_orders.get_mut(&tpid) {
                    v.retain(|&id| id != order_id);
                }
            }
        }
        if let Some(bts) = self.order_bintags.remove(&order_id) {
            for bt in bts {
                if let Some(v) = self.bintag_to_orders.get_mut(&bt) {
                    v.retain(|&id| id != order_id);
                }
            }
        }
    }
}

impl Default for InvertedIndexes {
    fn default() -> Self {
        Self::new()
    }
}
